### PR TITLE
Add console extension manfiest for ConsoleExternalLogLinks CRD

### PR DESCRIPTION
**Pull request by rhamilto**
Resolves https://jira.coreos.com/browse/CONSOLE-1547

### CODE REVIEW

File: manifests/00-crd-extension-console-pod-log-link.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+          required:
+          - text
+          - hrefTemplate
+          properties:
+            text:
+              type: string
+              description: Text of the link
+            hrefTemplate:
+              type: string
+              description: Absolute secure URL including pod name or uid, container name, namespace, and labels variables for the pod log link (must use https).  e.g., https://example.com/logs?name=\${name}&container=\${container}&namespace=\${namespace}&labels=\${labels}
```
Comment (spadgett) [inline manifests/00-crd-extension-console-pod-log-link.yaml]: We'll want to document specifically all the replacements we support and the exact format

File: manifests/00-crd-extension-console-pod-log-link.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: consolepodloglinks.console.openshift.io
+  annotations:
+    displayName: ConsolePodLogLinks
```
Comment (benjaminapetersen) [inline manifests/00-crd-extension-console-pod-log-link.yaml]: And per our slack, considering a name change to make more generic

**Comment by benjaminapetersen** [comment]
/hold 

Note that `ci/prow/e2e-aws` will fail unless we bump the `rbac` test again.  

Since we are holding off on this PR for design, We will likely have the CRD proxy for the console server done by then, so shouldn't need to do this.

**Comment by rhamilto** [comment]
Updated.  @spadgett, PTAL.

**Comment by benjaminapetersen** [comment]
Needs an update like this one for the tests to pass:
https://github.com/openshift/origin/pull/23231
Working on that now.

File: manifests/03-rbac-role-cluster-extensions.yaml [hunk 3]
```diff
@@ -11,6 +11,7 @@ rules:
   - consolelinks
   - consoleclidownloads
   - consolenotifications
+  - consoleloglinks
```
Comment (benjaminapetersen) [inline manifests/03-rbac-role-cluster-extensions.yaml]: `consoleexternalloglinks`

https://github.com/openshift/origin/pull/23382/files#diff-7a19fbef2a463f9e30c55aa61cf6f5ffR119

File: manifests/00-crd-extension-console-log-link.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: consoleloglinks.console.openshift.io
```
Comment (benjaminapetersen) [inline manifests/00-crd-extension-console-log-link.yaml]: `ConsoleExternalLogLink` now 💩

File: manifests/00-crd-extension-console-log-link.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+              type: string
+              description: Text of the link
+            hrefTemplate:
+              type: string
+              description: >-
+                Absolute secure URL (must use https) for the log link including variables to be replaced.
+                Supported variables include ${resourceName}, ${resourceUID}, ${containerName}, ${resourceNamespace}, and ${podLabels} and must match exactly.
+                e.g., https://example.com/logs?resourceName=\${resourceName}&containerName=\${containerName}&resourceNamespace=\${resourceNamespace}&podLabels=\${podLabels}
+              pattern: '^https://'
+            namespaceFilter:
```
Comment (benjaminapetersen) [inline manifests/00-crd-extension-console-log-link.yaml]: Trying to decide if we want a pattern on this. For example:
`^openshift-` vs `/^openshift-/` vs `/^openshift-/g` 

We are using `string.search(regex)`, so `^openshift-` will break, but the other two will work. [This stack overflow](https://stackoverflow.com/questions/172303/is-there-a-regular-expression-to-detect-a-valid-regular-expression) says that this is a perfect regex to check for a regex:

`/^((?:(?:[^?+*{}()[\]\\|]+|\\.|\[(?:\^?\\.|\^[^\\]|[^\\^])(?:[^\]\\]+|\\.)*\]|\((?:\?[:=!]|\?<[=!]|\?>)?(?1)??\)|\(\?(?:R|[+-]?\d+)\))(?:(?:[?+*]|\{\d+(?:,\d*)?\})[?+]?)?|\|)*)$/
`
I kid about including that 😄 

To make cross reference easy: https://github.com/openshift/console/pull/1892/files#r304608136

We may want to do something, however? At least check for leading `/`?  I don't think we should necessarily try to be comprehensive, that sounds like bugs.

**Comment by spadgett** [comment]
@rhamilto we should match the descriptions here to what's in openshift/api since it's what's added to the OpenAPI doc (oc explain, API Explorer, etc.)

Note there is one minor change: https://github.com/openshift/api/pull/385

File: manifests/00-crd-extension-console-log-link.yaml (hunk 5 above)
Comment (spadgett) [reply manifests/00-crd-extension-console-log-link.yaml]: Per discussion in https://github.com/openshift/api/pull/385, leading `/` is not necessary. I'm not sure what validation is really practical here. I think this is reasonable for now, and we can improve validation as a bug fix if needed.

Comment (benjaminapetersen) [reply manifests/00-crd-extension-console-log-link.yaml]: Works for me.

File: manifests/00-crd-extension-console-log-link.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        spec:
+          type: object
+          description: Represents console external log link customizations spec
+          required:
+          - text
+          - hrefTemplate
+          properties:
+            text:
+              type: string
+              description: Text of the link
```
Comment (benjaminapetersen) [inline manifests/00-crd-extension-console-log-link.yaml]: `Text is the display text of the link`

https://github.com/openshift/api/pull/380/files#diff-9754f97c9c90951cc7df5ac76b6eea86R20

**Comment by benjaminapetersen** [comment]
@rhamilto want to retitle PR for `ConsoleExternalLogLink`? (just for future reference parity)

File: manifests/00-crd-extension-console-log-link.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            hrefTemplate:
```
Comment (benjaminapetersen) [inline manifests/00-crd-extension-console-log-link.yaml]: `hrefTemplate` description is pretty close, though not exact.  Do we want to just copy/paste to make maintenance over time easy?

**Comment by rhamilto** [comment]
Descriptions updated to align with api.  Note I had to make some formatting adjustments to make the text yaml-friendly (removed colon after `Supported variables are` and changed `-` bullet to `*`.
@spadgett, @benjaminapetersen

**Comment by spadgett** [comment]
Feel free to remove the hold when https://github.com/openshift/origin/pull/23382 lands

**Pull request merged**